config/runtime.exs

```diff
@@ -4,6 +4,27 @@ filter_nil_kv_pairs = fn pairs when is_list(pairs) ->
   Enum.filter(pairs, fn {_k, v} -> v !== nil end)
 end
 
+detect_ip_version = fn host ->
+  host = String.to_charlist(host)
+
+  cond do
+    match?({:ok, _}, :inet6_tcp.getaddr(host)) -> {:ok, :inet6}
+    match?({:ok, _}, :inet.gethostbyname(host)) -> {:ok, :inet}
+    true -> {:error, :nxdomain}
+  end
+end
+
+socket_options_for_host = fn
+  host when is_binary(host) ->
+    case detect_ip_version.(host) do
+      {:ok, ip_version} -> [ip_version]
+      {:error, reason} -> raise "Failed to detect IP version: #{reason}"
+    end
+
+  _host ->
+    []
+end
```

Comment:
For what is worth, config/runtime.exs can access modules defined in your application (as long as it doesn't call any process, as the app did not start yet, so you can move these functions to a module if you want (especially if you want to write some tests).